architecture/2025-06-compute-extensibility-feature-spec.md

```diff
@@ -0,0 +1,1136 @@
+* Ingress controller customization is enabled through the `Radius.Compute/gateways@2025-05-01-preview` resource type, where platform engineers can implement the ingress controller solutions of their choice using Recipes. Contour installation is no longer a hard dependency for Radius, allowing users to use alternative gateway solutions like NGINX Ingress Controller or Traefik.
+* Recipe Packs are introduced as a new concept to bundle and distribute related recipes together, making it easier to manage and share them across different environments. Recipe Packs are modeled as Radius resources, specifically `Radius.Core/recipePacks@2025-05-01-preview`, and can be referenced by environments.
+* Resource types are renamed to follow the new Radius naming convention:
+    ```
+    `Applications.Core/environments` --> `Radius.Core/environments`
+    `Applications.Core/applications` --> `Radius.Core/applications`
+    `Radius.Core/recipePacks` --> `Radius.Core/recipePacks` (new `recipePacks` resource added to `Core` namespace)
+    `Applications.Core/containers` --> `Radius.Compute/containers`
+    `Applications.Core/gateways` --> `Radius.Compute/gateways`
+    `Applications.Core/volumes` --> `Radius.Storage/volumes`
+    `Applications.Core/secretStores` --> `Radius.Security/secrets`
+    `Applications.Datastores/sqlDatabases` --> `Radius.Data/sqlDatabases`
+    `Applications.Datastores/mongoDatabases` --> `Radius.Data/mongoDatabases`
+    `Applications.Datastores/redisCaches` --> `Radius.Data/redisCaches`
+    `Applications.Messaging/rabbitmqQueues` --> `Radius.Data/rabbitMQQueues`
+    `Applications.Dapr/*` --> `Radius.Dapr/*`
+    `Radius.Config/*` --> `Radius.Config/*` (new namespace to support future configurations like `terraformSettings`)
+    `Radius.Resources/*` --> `Radius.Resources/*` (user defined custom resource types)
+    ```
+
+## Key investments
+<!-- List the features required to enable this scenario(s). -->
+
+### Feature 1: RRT Implementation of Core Application Model Types
+<!-- One or two sentence summary -->
+Re-implement existing core types (`Applications.Core/environments`, `Applications.Core/applications`, `Applications.Core/containers`, `Applications.Core/gateways`, `Applications.Core/secrets`, `Applications.Core/volumes`) as Radius Resource Types (RRT) with new, versioned resource type names (i.e. `Radius.Core/environments`, `Radius.Core/applications`, `Radius.Compute/containers`, `Radius.Compute/gateways`, `Radius.Security/secrets`, `Radius.Storage/volumes`). These RRTs will form the basis of the extensible application model. These new core types must support Connections and may be modified by platform engineers to allow for the configuration of platform-specific capabilities (e.g. confidential containers) in the resource types themselves.
```

Comment:
If we implement `Applications.Core/environments` and `Applications.Core/applications` as RRTs, would customers be able to modify them and create recipes for them? How would we handle this in the CLI and API? Environments and Applications are primary concepts in the CLI, graph, and dashboard, and have expectations about behavior and validation. What would happen if someone registered a new version of those types?

I think there could be value in allowing customers to add on or extend the Environment and Application to track data that is important to them, but that seems out of scope for this feature spec. What do you think?

Reply:
updated, please take another look @brooke-hamilton